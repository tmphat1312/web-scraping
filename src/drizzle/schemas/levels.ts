import { integer, sqliteTable, text } from 'drizzle-orm/sqlite-core';

export const levels = sqliteTable('levels', {
  id: integer('id').primaryKey({ autoIncrement: true }),
  jp_name: text('jp_name').notNull().unique(),
  en_name: text('en_name').notNull().unique(),
  min_value: integer('min_value').notNull().unique(),
  max_value: integer('max_value').notNull().unique(),
});
