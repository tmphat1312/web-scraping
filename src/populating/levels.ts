import { fromURL } from 'cheerio';

scrapeLevels().then(console.log);

export async function scrapeLevels() {
  let $ = await fromURL(`https://www.wanikani.com/kanji`);
  return $.extract({
    levels: [
      {
        selector: `.sitemap__group-header`,
        value: (el) => {
          let [jp_name, en_name] = $(el).text().split(' ');
          return {
            jp_name,
            en_name,
            min_value: $(el).next().find('a').first().text(),
            max_value: $(el).next().find('a').last().text(),
          };
        },
      },
    ],
  }).levels;
}
