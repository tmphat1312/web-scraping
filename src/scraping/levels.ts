import { fromURL } from 'cheerio';

export async function scrapeLevels() {
  let $ = await fromURL(`https://www.wanikani.com/kanji`);
  return $.extract({
    levels: [
      {
        selector: `.sitemap__group-header`,
        value: (el) => {
          let [jaName, enName] = $(el).text().split(' ');
          return {
            jaName,
            enName,
            minValue: Number($(el).next().find('a').first().text()),
            maxValue: Number($(el).next().find('a').last().text()),
          };
        },
      },
    ],
  }).levels;
}
