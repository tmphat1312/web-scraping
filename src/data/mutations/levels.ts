import { InferInsertModel } from 'drizzle-orm';

import { db } from '../client';
import { levels } from '../schemas/levels';

export type Level = InferInsertModel<typeof levels>;

export async function insertLevels(data: Level[]) {
  return db
    .insert(levels)
    .values(data)
    .returning({
      id: levels.id,
    })
    .onConflictDoNothing();
}

export async function truncateLevels() {
  return db.delete(levels);
}
