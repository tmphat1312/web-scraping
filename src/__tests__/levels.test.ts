import { scrapeLevels } from '../populating/scripts/levels';

describe('scrape levels', () => {
  it('should be the same levels', async () => {
    let levels = await scrapeLevels();
    expect(levels).toMatchSnapshot();
  });
});
