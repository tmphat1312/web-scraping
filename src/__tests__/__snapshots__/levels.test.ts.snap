// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scrape levels > should be the same levels 1`] = `
[
  {
    "en_name": "PLEASANT",
    "jp_name": "快",
    "max_value": "10",
    "min_value": "01",
  },
  {
    "en_name": "PAINFUL",
    "jp_name": "苦",
    "max_value": "20",
    "min_value": "11",
  },
  {
    "en_name": "DEATH",
    "jp_name": "死",
    "max_value": "30",
    "min_value": "21",
  },
  {
    "en_name": "HELL",
    "jp_name": "地獄",
    "max_value": "40",
    "min_value": "31",
  },
  {
    "en_name": "PARADISE",
    "jp_name": "天国",
    "max_value": "50",
    "min_value": "41",
  },
  {
    "en_name": "REALITY",
    "jp_name": "現実",
    "max_value": "60",
    "min_value": "51",
  },
]
`;
